contracts/Controller.sol

```diff
@@ -522,13 +544,75 @@ contract Controller is ReentrancyGuard, Ownable {
      * @notice exercise option
      * @param _args ExerciseArgs structure
      */
-    // function _exercise(Actions.ExerciseArgs memory _args) internal {}
+    function _exercise(Actions.ExerciseArgs memory _args) internal {
+        OtokenInterface otoken = OtokenInterface(_args.otoken);
+
+        require(now > otoken.expiryTimestamp(), "Controller: can not exercise un-expired otoken");
+
+        require(isPriceFinalized(_args.otoken), "Controller: otoken underlying asset price is not finalized yet");
+
+        uint256 payout = _getPayout(_args.otoken, _args.amount);
+
+        otoken.burnOtoken(msg.sender, _args.amount);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        marginPool.transferToUser(otoken.collateralAsset(), _args.receiver, payout);
+
+        emit Exercise(_args.otoken, msg.sender, _args.receiver, otoken.collateralAsset(), _args.amount, payout);
+    }
 
     /**
      * @notice settle vault option
      * @param _args SettleVaultArgs structure
      */
-    // function _settleVault(Actions.SettleVaultArgs memory _args) internal {}
+    function _settleVault(Actions.SettleVaultArgs memory _args)
+        internal
+        isAuthorized(msg.sender, _args.owner)
+        returns (MarginAccount.Vault memory)
+    {
+        require(checkVaultId(_args.owner, _args.vaultId), "Controller: invalid vault id");
+
+        MarginAccount.Vault memory vault = vaults[_args.owner][_args.vaultId];
+
+        require(isNotEmpty(vault.shortOtokens), "Controller: can not settle a vault with no otoken minted");
+
+        OtokenInterface shortOtoken = OtokenInterface(vault.shortOtokens[0]);
+
+        require(now > shortOtoken.expiryTimestamp(), "Controller: can not settle vault with un-expired otoken");
+        require(
+            isPriceFinalized(address(shortOtoken)),
+            "Controller: otoken underlying asset price is not finalized yet"
+        );
+
+        address calculatorModule = AddressBookInterface(addressBook).getMarginCalculator();
+        MarginCalculatorInterface calculator = MarginCalculatorInterface(calculatorModule);
+
+        (uint256 payout, ) = calculator.getExcessCollateral(vault);
+
+        address marginPoolModule = AddressBookInterface(addressBook).getMarginPool();
+        MarginPoolInterface marginPool = MarginPoolInterface(marginPoolModule);
+
+        if (isNotEmpty(vault.longOtokens)) {
+            OtokenInterface longOtoken = OtokenInterface(vault.longOtokens[0]);
+
+            longOtoken.burnOtoken(marginPoolModule, vault.longAmounts[0]);
+        }
+
+        vaults[_args.owner][_args.vaultId]._clearVault();
```

Comment:
```suggestion
        vaults[_args.owner][_args.vaultId]._clearVault();
        if (isNotEmpty(vault.longOtokens)) {
            OtokenInterface longOtoken = OtokenInterface(vault.longOtokens[0]);

            longOtoken.burnOtoken(marginPoolModule, vault.longAmounts[0]);
        }
```
Better update state first before external calls

Reply:
Yeah, but if I clear vault first, then `vault.longOtokens` will be empty, also we are making a call to a long otoken in a vault that is depositing through the protocol, so all long otoken that are there are whitelisted and verified.